torch/lib/THCUNN/SpatialDepthwiseConvolution.cu

```diff
@@ -0,0 +1,191 @@
+// updateOutput, updateGradInput Kernels ported from Sergey Zagoruyko's pyinn, which itself was a
+// port from Caffe
+
+#include "THCUNN.h"
+#include "THCDeviceTensor.cuh"
+#include "THCDeviceTensorUtils.cuh"
+#include "THCNumerics.cuh"
+#include "THCReduceApplyUtils.cuh"
+#include "THCSortUtils.cuh"
+#include "THCTensorMathReduce.cuh"
+#include "SharedMem.cuh"
+#include "common.h"
+
+template <typename T, typename IndexType>
+__global__ void spatialDepthwiseConvolutionUpdateOutput(
+    const THCDeviceTensor<T, 4> input,
+    THCDeviceTensor<T, 4> output,
+    const THCDeviceTensor<T, 4> weight,
+    const THCDeviceTensor<T, 1> bias,
+    bool biasEnabled,
+    IndexType totalElements,
+    const int outputChannels,
+    const int depthwiseMultiplier,
+    const int inputWidth, const int inputHeight,
+    const int outputWidth, const int outputHeight,
+    const int kernelWidth, const int kernelHeight,
+    const int strideWidth, const int strideHeight,
+    const int padWidth, const int padHeight,
+    const int dilationWidth, const int dilationHeight)
+{
+  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < totalElements;
+       linearIndex += gridDim.x * blockDim.x) {
+
+    const int n = linearIndex / outputChannels / outputHeight / outputWidth;
+    const int c = (linearIndex / outputHeight / outputWidth) % outputChannels;
+    const int h = (linearIndex / outputWidth) % outputHeight;
+    const int w = linearIndex % outputWidth;
```

Comment:
I've seen integer divison and modulo being ridiculously expensive in CUDA, so it's best to avoid it. You can take a look at my Softmax PR to see how I used a 2d grid to tile the blocks, which let me avoid this indexing at all and made the kernel a few times faster.

---

Also a suggestion: I usually find the kernels much easier to read when you pre-compute all the strides outside as `const int` and rewrite the indexing to look more like this:
```cpp
// outside of the loop
const int widthSize = outputWidth;
const int heightStride = widthSize;
const int heightSize = outputHeight;
const int channelStride = heightSize * heightStride;
const int channelSize = outputChannels;
const int batchStride = channelSize * channelStride;
// inside the loop
const int w = linearIndex % widthSize;
const int h = (linearIndex / heightStride) % heightSize;
const int c = (linearIndex / channelStride) % channelSize;
const int n = linearIndex / batchStride;
```

Reply:
Also addressed by @ngimel above. I find that these kernels are pretty ugly no matter what.